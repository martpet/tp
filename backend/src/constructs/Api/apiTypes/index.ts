export * from './EndpointsCookies';
export * from './EndpointsEnvironments';
export * from './EndpointsHeaders';
export * from './EndpointsOptions';
export * from './EndpointsQueryStrings';
export * from './OauthCookieProps';
export * from './OauthTokens';
