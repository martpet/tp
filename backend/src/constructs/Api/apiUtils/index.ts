export * from './cookieName';
export * from './errorResponse';
export * from './handlerWrapper';
export * from './itResolvesWithApiError';
