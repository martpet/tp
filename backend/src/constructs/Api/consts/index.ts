export * from './authorizationHeader';
export * from './defaultCachePolicyProps';
export * from './defaultSecurityHeadersBehavior';
export * from './lambdaEdgeViewerEvent';
