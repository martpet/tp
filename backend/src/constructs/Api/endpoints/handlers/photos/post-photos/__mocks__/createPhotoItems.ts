export const createPhotoItems = vi
  .fn()
  .mockName('createPhotoItems')
  .mockResolvedValue(undefined);
