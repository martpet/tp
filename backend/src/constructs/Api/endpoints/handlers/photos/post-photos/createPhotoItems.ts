import {
  BatchWriteCommand,
  DynamoDBClient,
  DynamoDBDocument,
  PhotosTableItem,
  photosTableOptions,
  PostPhotosRequest,
  unmarshall,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocument.from(ddbClient);

type Props = {
  requestData: PostPhotosRequest;
  sub: string;
};

export const createPhotoItems = async ({ requestData, sub }: Props) => {
  const { tableName } = photosTableOptions;

  async function batchWrite(photoItems: PhotosTableItem[]) {
    const batchWriteCommand = new BatchWriteCommand({
      RequestItems: { [tableName]: photoItems.map((Item) => ({ PutRequest: { Item } })) },
    });
    const { UnprocessedItems } = await ddbDocClient.send(batchWriteCommand);

    if (UnprocessedItems) {
      const nextItems: PhotosTableItem[] = [];
      UnprocessedItems[tableName]?.forEach(({ PutRequest }) => {
        if (PutRequest?.Item) {
          nextItems.push(unmarshall(PutRequest.Item) as PhotosTableItem);
        }
      });
      if (nextItems.length) {
        await batchWrite(nextItems);
      }
    }
  }

  const batchSize = 25;
  const requests = [];

  for (let i = 0; i < requestData.length; i += batchSize) {
    const chunk = requestData.slice(i, i + batchSize);
    const photoItems = chunk.map((item) => ({
      ...item,
      userId: sub,
      createdAt: Date.now(),
    }));
    requests.push(batchWrite(photoItems));
  }

  await Promise.all(requests);
};
