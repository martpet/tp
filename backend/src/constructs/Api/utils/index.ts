export * from './cookieName';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './lambdaEdgeErrorResponse';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
