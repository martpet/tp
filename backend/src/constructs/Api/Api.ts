import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

import { Auth, Photos, Tables, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getEnvName } from '~/utils';

import { createDistroBehaviors } from './endpoints/authEdgeFunction/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  tables: Tables;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, tables, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getEnvName(this);
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'HttpApi');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
    });

    createRoutes({ scope, api, auth, tables, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      tables,
      distribution,
      origin,
    });

    new ARecord(this, 'Alias', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
