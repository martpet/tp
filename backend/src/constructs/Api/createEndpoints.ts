import { HttpApi, HttpMethod } from '@aws-cdk/aws-apigatewayv2-alpha';
import { HttpLambdaIntegration } from '@aws-cdk/aws-apigatewayv2-integrations-alpha';
import {
  CachePolicy,
  Distribution,
  OriginRequestCookieBehavior,
  OriginRequestHeaderBehavior,
  OriginRequestPolicy,
  OriginRequestPolicyProps,
  OriginRequestQueryStringBehavior,
} from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { NodejsFunctionProps } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { Writable } from 'type-fest';

import { capitalize, createNodejsFunction } from '~/utils';

import { EndpointsOptions, MethodOptions } from './apiTypes';

type CreateEndpointProps = {
  scope: Construct;
  distribution: Distribution;
  api: HttpApi;
  origin: HttpOrigin;
  options: EndpointsOptions;
};

export const createEndpoints = ({
  scope,
  distribution,
  api,
  origin,
  options,
}: CreateEndpointProps) => {
  Object.entries(options).forEach(
    ([path, { methods, cookies, headers, queryStrings }]) => {
      addDistroBehavior({
        scope,
        distribution,
        origin,
        path,
        headers,
        cookies,
        queryStrings,
      });

      Object.entries(methods).forEach(([method, methodOptions]) => {
        addApiRoute({ scope, api, path, method, methodOptions });
      });
    }
  );
};

type AddDistroBehaviorProps = {
  scope: Construct;
  distribution: Distribution;
  origin: HttpOrigin;
  path: string;
  headers?: string[];
  cookies?: string[];
  queryStrings?: Readonly<string[]>;
};

function addDistroBehavior({
  scope,
  distribution,
  origin,
  path,
  cookies,
  headers,
  queryStrings,
}: AddDistroBehaviorProps) {
  const pathName = path.replace('/', '');
  const policyProps: Writable<OriginRequestPolicyProps> = {};

  if (headers) {
    policyProps.headerBehavior = OriginRequestHeaderBehavior.allowList(...headers);
  }

  if (cookies) {
    policyProps.cookieBehavior = OriginRequestCookieBehavior.allowList(...cookies);
  }

  if (queryStrings) {
    policyProps.queryStringBehavior = OriginRequestQueryStringBehavior.allowList(
      ...queryStrings
    );
  }

  distribution.addBehavior(path, origin, {
    cachePolicy: CachePolicy.CACHING_DISABLED,
    originRequestPolicy: new OriginRequestPolicy(scope, pathName, policyProps),
  });
}

type AddApiRouteProps = {
  scope: Construct;
  api: HttpApi;
  path: string;
  method: string;
  methodOptions: MethodOptions<unknown>;
};

function addApiRoute({ scope, api, path, method, methodOptions }: AddApiRouteProps) {
  const pathName = path.replace('/', '');
  const methodName = method.toLowerCase();
  const fileName = `${methodName}-${pathName}`;
  const id = `${capitalize(methodName)}${capitalize(fileName)}`;
  const { environment, cb } = methodOptions;
  const handlerProps: Writable<NodejsFunctionProps> = {
    entry: `${__dirname}/handlers/${pathName}/${fileName}/${fileName}.ts`,
    environment,
  };

  const handler = createNodejsFunction(scope, `${id}Handler`, handlerProps);

  if (cb) {
    cb(handler);
  }

  api.addRoutes({
    path,
    methods: [method as HttpMethod],
    integration: new HttpLambdaIntegration(`${id}Integration`, handler),
  });
}
