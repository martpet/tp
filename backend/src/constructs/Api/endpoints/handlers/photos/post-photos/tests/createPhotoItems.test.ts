import { BatchWriteCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import { photosTableOptions } from 'lambda-layer';

import { itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';

import { createPhotoItems } from '../createPhotoItems';

vi.mock('@aws-sdk/util-dynamodb');

const ddbMock = mockClient(DynamoDBDocumentClient);

const args = [
  {
    requestData: [{ fingerprint: 'dummyFingerprint' }],
    sub: 'dummySub',
  },
] as Parameters<typeof createPhotoItems>;

beforeEach(() => {
  ddbMock.reset();
  ddbMock.on(BatchWriteCommand).resolves({});
});

describe('createPhotoItems', () => {
  itSendsAwsCommand(BatchWriteCommand, ddbMock, createPhotoItems, args);
  itResolves(createPhotoItems, args);

  describe('when called with 26 items', async () => {
    const argsClone = structuredClone(args);
    argsClone[0].requestData = Array(26);

    it('sends "BatchWriteCommand" 2 times', async () => {
      await createPhotoItems(...argsClone);
      expect(ddbMock.commandCalls(BatchWriteCommand).length).toBe(2);
    });
  });

  describe('when "BatchWriteCommand" response has "UnprocessedItems"', () => {
    beforeEach(() => {
      ddbMock
        .on(BatchWriteCommand)
        .resolvesOnce({
          UnprocessedItems: {
            [photosTableOptions.tableName]: [
              { PutRequest: { Item: { dummyUnprocessedKey: 'dummyUnprocessedValue' } } },
            ],
          },
        })
        .resolvesOnce({});
    });

    it('sends "BatchGetCommand" a second time with correct args', async () => {
      await createPhotoItems(...args);
      expect(ddbMock.commandCalls(BatchWriteCommand)[1].args[0].input).toMatchSnapshot();
    });
  });
});
