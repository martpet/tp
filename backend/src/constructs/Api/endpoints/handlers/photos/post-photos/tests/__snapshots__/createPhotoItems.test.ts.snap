// Vitest Snapshot v1

exports[`createPhotoItems > resolves with a correct value 1`] = `undefined`;

exports[`createPhotoItems > sends "BatchWriteCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "RequestItems": {
    "TP-photos": [
      {
        "PutRequest": {
          "Item": {
            "createdAt": 1640995200000,
            "fingerprint": "dummyFingerprint",
            "userId": "dummySub",
          },
        },
      },
    ],
  },
}
`;

exports[`createPhotoItems > when "BatchWriteCommand" response has "UnprocessedItems" > sends "BatchGetCommand" a second time with correct args 1`] = `
{
  "RequestItems": {
    "TP-photos": [
      {
        "PutRequest": {
          "Item": "dummyUnmarshalledItems",
        },
      },
    ],
  },
}
`;
